shopfloor/actions/stock.py

```diff
@@ -10,6 +13,98 @@ class StockAction(Component):
     _inherit = "shopfloor.process.action"
     _usage = "stock"
 
+    def _create_return_move__get_max_qty(self, origin_move):
+        """Returns the max returneable qty."""
+        # The max returnable qty is the sent qty minus the already returned qties
+        quantity = origin_move.product_qty
+        for move in origin_move.move_dest_ids:
+            if (
+                move.origin_returned_move_id
+                and move.origin_returned_move_id != origin_move
+            ):
+                continue
+            if move.state in ("partially_available", "assigned"):
+                quantity -= sum(move.move_line_ids.mapped("product_qty"))
+            elif move.state in ("done"):
+                quantity -= move.product_qty
+        return float_round(
+            quantity, precision_rounding=origin_move.product_id.uom_id.rounding
+        )
+
+    def _create_return_move__get_vals(self, return_picking, origin_move):
+        product = origin_move.product_id
+        return_type = return_picking.picking_type_id
+        return {
+            "product_id": product.id,
+            "product_uom": product.uom_id.id,
+            "picking_id": return_picking.id,
+            "state": "draft",
+            "date": fields.Datetime.now(),
+            "location_id": return_picking.location_id.id,
+            "location_dest_id": return_picking.location_dest_id.id,
+            "picking_type_id": return_type.id,
+            "warehouse_id": return_type.warehouse_id.id,
+            "origin_returned_move_id": origin_move.id,
+            "procure_method": "make_to_stock",
+        }
+
+    def _create_return_move__link_to_origin(self, return_move, origin_move):
+        move_orig_to_link = origin_move.move_dest_ids.mapped("returned_move_ids")
+        move_orig_to_link |= origin_move
```

Comment:
Questions, the first parameter is the returned move?
If the `origin_move` is the move going to Customers location (so the move to return), its `move_dest_ids` field should list the returned moves already (Customers -> Stock)? Why are we reading `returned_move_ids` from there?
A docstring explaining what we want to achieve here is welcome.
```suggestion
    def _create_return_move__link_to_origin(self, returned_move, origin_move):
        move_orig_to_link = origin_move.move_dest_ids.mapped("returned_move_ids")
        move_orig_to_link |= origin_move
```

Reply:
I see what you mean.
Origin move is the returned move. Its content has been returned.
It has generated a return move, which is return_move.
I can rename it "new_move" if you think it is clearer.

Reply:
Baaah, there's notes everywhere in this file about what it does, and where this has been copied from.
However, I can rename `return_move` to `new_move` if you think it's clearer.